R/fabricate.R

```diff
@@ -35,215 +52,667 @@
 #' # Draw a two-level hierarchical dataset
 #' # containing cities within regions
 #' df <- fabricate(
-#'  regions = level(N = 5),
-#'  cities = level(N = 2, pollution = rnorm(N, mean = 5)))
+#'  regions = add_level(N = 5),
+#'  cities = add_level(N = 2, pollution = rnorm(N, mean = 5)))
 #' head(df)
 #'
 #' # Start with existing data and add variables to hierarchical data
-#' # note: do not provide N when adding variables to an existing level
+#' # at levels which are already present in the existing data.
+#' # Note: do not provide N when adding variables to an existing level
 #' df <- fabricate(
 #'   data = df,
-#'   regions = level(watershed = sample(c(0, 1), N, replace = TRUE)),
-#'   cities = level(runoff = rnorm(N))
+#'   regions = modify_level(watershed = sample(c(0, 1), N, replace = TRUE)),
+#'   cities = modify_level(runoff = rnorm(N))
 #' )
 #'
-#' @importFrom rlang quos quo_name eval_tidy lang_name lang_modify lang_args is_lang get_expr
+#' @importFrom rlang quos quo_name eval_tidy lang_name lang_modify lang_args
+#' is_lang get_expr
 #'
 #' @export
-fabricate <-
-  function(data,
-           N,
-           ID_label,
-           ...) {
-    options <- quos(...)
-
-    all_levels <- check_all_levels(options)
-
-    if (!missing(data) && !"data.frame"  %in% class(data)) {
-      if(is.null(dim(data))) {
-          stop(
-            "User provided data must be a data frame. Provided data was low dimensional."
-          )
-      }
-      if(!"data" %in% names(sys.call())) {
-        stop(
-          "The data argument must be a data object. The argument call, ", deparse(substitute(data)), ", was not a data object (e.g. a data.frame, tibble, sf object, or convertible matrix)."
-        )
-      }
-      tryCatch({
-        data = as.data.frame(data)
-      }, error=function(e) {
-        stop(
-          "User provided data could not convert to a data frame."
-        )
-      })
+fabricate <- function(data = NULL, ..., N = NULL, ID_label = NULL)
+{
+  # Store all data generation arguments in a quosure for future evaluation
+  # A quosure contains unevaluated formulae and function calls.
+  data_arguments = quos(...)
+
+  # Fabricatr expects either a single-level function call
+  # or a series of level calls. You can't mix and match.
+  # This helper function will be TRUE if calls are all levels, FALSE
+  # if there are no calls or they are not levels.
+  all_levels = check_all_levels(data_arguments)
+
+  # User must provide exactly one of:
+  # 1) One or more level calls (with or without importing their own data)
+  # 2) Import their own data and do not involve level calls
+  # 3) Provide an N without importing their own data
+  if(sum((!is.null(data) & !missing(data) & !all_levels),
```

Comment:
missing should be to the left of the is.null, also use &&s here to short circuit it. Also the comment is better than the error message below.